---
title: "Exploratory Data Analysis in R"
author: "James Sacco"
date: "3/28/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

###STATISTICS FOR GENOMIC DATA SCIENCE: MODULE 1 
##EXPLORATORY DATA ANALYSIS IN R

reference: http://jtleek.com/genstats/inst/doc/01_10_exploratory-analysis.html

##PART 1

use the below code if you need to download a CRAN package from BioConductor

source("http://bioconductor.org/biocLite.R")
biocLite("pkgname")

```{r}
rm(list=ls())
```

load libraries
```{r}
library(gplots)
library(devtools)
library(Biobase)
library("RSkittleBrewer", lib.loc="/Library/Frameworks/R.framework/Versions/3.3/Resources/library")
library(org.Hs.eg.db)
library(AnnotationDbi)
library(dplyr)
```

make plots pretty

```{r}
trop = RSkittleBrewer("tropical")
palette(trop)

par(pch=19)
```
```

load data

```{r}
con = url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bodymap_eset.RData")
load(file=con)
close(con)
bm = bodymap.eset
pdata = pData(bm)
edata = exprs(bm)
fdata = fData(bm)
ls()
```

use the above code chunk to start each Part (1,2,3)

summarize data
```{r}
table(pdata$gender)
table(pdata$gender, pdata$race)

summary(edata)
```
missing values
Use option useNA to include NA's in table
```{r}
table(pdata$age, useNA = "ifany")
```
is.na checks for NA values
```{r}
table(is.na(pdata$age))
```
Check for other common missing names
```{r}
sum(pdata$age == " ")
```
remove NAs in age column
```{r}
sum(pdata$age == " ", na.rm=TRUE)
```
check genomic data for NAs
```{r}
sum(is.na(edata))
```
make the distribution of NAs by gene
```{r}
gene_na = rowSums(is.na(edata))
table(gene_na)
```
Make the distribution of NAs by samples
```{r}
sample_na = rowSums(is.na(edata))
table(sample_na)
```
Check that dimensions match up

The number of rows of the feature data should match the number of rows of the genomic data 
(both are the number of genes). The number of rows of the phenotype data should match the 
number of columns of the genomic data (both are the number of samples).

dimenensions
```{r}
dim(fdata) #features data
dim(pdata) #phenotype data
dim(edata) #expression data
```

##PART 2
Look at overall distributions

```{r}
rm(list=ls())
library(gplots)
library(devtools)
library(Biobase)
library("RSkittleBrewer", lib.loc="/Library/Frameworks/R.framework/Versions/3.3/Resources/library")
library(org.Hs.eg.db)
library(AnnotationDbi)
library(dplyr)


```


```{r}
con = url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bodymap_eset.RData")
load(file=con)
close(con)
bm = bodymap.eset
pdata = pData(bm)
edata = exprs(bm)
fdata = fData(bm)
ls()
```


boxplot
```{r}
boxplot(edata[,1])
```

The above is a useless plot; most of values are near zero, and there are many outliers

```{r}
boxplot(log2(edata+1), col=2, range=0)
```

Histograms

look at this, sample by sample, with histograms

show plots side by side on the screan: 1 row and 2 columns of plots
```{r}
par(mfrow=c(1,2)) 
```

almost all values are equal to 0
```{r}
hist(log2(edata[,1]+1),col=2) 
hist(log2(edata[,2]+1),col=2)
```

change display for plots 1 row and 1 column of plots
```{r}
par(mfrow=c(1,1)) 
```

density plot of sample 1

log transform and density of 1st sample
```{r}
plot(density(log2(edata[,1]+1)), col=2)
```

add density plot of sample 2

layer other values on top with the lines command
```{r}
plot(density(log2(edata[,2]+1)), col=2)

lines(density(log2(edata[,1]+1)), col=2)
```



A very common task is to compare distributions of measurements (say, before normalization). You can do this with a qq-plot
```{r}
qqplot(log2(edata[,1]+1),log2(edata[,2]+1), col=3)
```

add 45 degree line to the qqplot with abline function
```{r}
qqplot(log2(edata[,1]+1),log2(edata[,2]+1), col=3)
abline(c(0,1))
```

make an M.A. Plot aka Bland-Altman Plot
```{r}
mm = log2(edata[,1]+1) - log2(edata[,2]+1)
aa = log2(edata[,1]+1) + log2(edata[,2]+1) 

plot(aa, mm, col=2)
```

for count based data, you need to remove low count features

(convert to dataframe to filter, and then convert to matrix to plot)

```{r}
edata = as.data.frame(edata) #convert to dataframe to filter


filt_edata = filter(edata, rowMeans(edata) > 1)
dim(filt_edata)

boxplot(as.matrix(log2(filt_edata +1)), col = 2) #convert to matrix to plot
```

##PART 3

CHECK FOR CONSISTENCY (or obvious data mixups). Compare data to metadata or other annotations

Here we are going to do a check to make sure that the men and women are correctly 
labeled by looking at expression on the Y chromosome. In general you might do several of this
type of check to confirm the data are correctly labeled.

Get the chromosomes for each gene using the feature data.

```{r}
aeid = as.character(fdata[,1])
chr = AnnotationDbi::select(org.Hs.eg.db,keys=aeid,keytype="ENSEMBL",columns="CHR")
head(chr)
```

Filter to the data on chromsome Y and sum up all the counts. A tricky issue is that some genes are 
annotated to multiple chromsomes. 

Here we take the first chromsome each is annotated to.

```{r}
dim(chr)

dim(edata)
```

Take non-duplicated chromsomes

```{r}
chr = chr[!duplicated(chr[,1]),]
```

confirm that the annotation is still in the right order
```{r}
all(chr[,1] == rownames(edata))
```

Select the chromosome Y samples

```{r}
edatay = dplyr::filter(edata,chr$CHR=="Y")
```
Note: if the above command does not work, you may need to use this work-around to eliminate the error, but you will get non-zero female values: edatay = dplyr::filter(as.data.frame(edata, as.character(chr$CHR=="Y"))) 


Males have Y chromosome expression as expected

```{r}
boxplot(colSums(edatay) ~ pdata$gender)
points(colSums(edatay) ~ jitter(as.numeric(pdata$gender)), 
       col=as.numeric(pdata$gender),
       p=19)
```


heatmaps

```{r}
ematrix = as.matrix(edata)[rowMeans(edata) > 10000,]
heatmap(ematrix)
```

change the coloring. make a color palette

```{r}
colramp = colorRampPalette(c(3,"blue",2))(9)
heatmap(ematrix,col=colramp)
```

remove automatic clustering

```{r}
heatmap(ematrix, col=colramp, Rowv = NA, Colv = NA)
```

Add color scale with heatmap.2 package. Make dendrogram disappear, scale the data by rows, and remove 
a tracing plot

```{r}
heatmap.2(ematrix,col=colramp,Rowv = NA, Colv = NA, dendrogram = "none", scale = "row", trace = "none")
```

session info

```{r, echo=FALSE}
devtools::session_info()
```

